Include Branch [dev|stable|beta] or pr$PullRequestID at end of version string

brtsai opened the pull request:
Fixes #917

I.E.: "1.4.0.0-fna + 1.5509.0-azure-ed1e7" (old)
=> "1.4.0.0-fna + 1.5509.0-azure-ed1e7-dev" (new)

brtsai commented:
Hi, I don't actually know how to test this change and would appreciate someone helping me with that. I'm also on mac and couldn't get Everest to build locally + I've never worked with Azure Pipelines before so definitely check that the branchname as derived from the env vars makes sense and is actually [dev|stable|beta] and not something like "main/dev" or the slash might have unintended consequences down the pipeline.

I rearranged the stuff suggested by copilot to make more readable sense (copilot suggested "1.$BuildNumber.0-$BranchName-$SourceVersionShort" but I thought branchname in the middle and not at the end would look weird while likely messing up existing version string parsing logic in [Everest.cs](https://github.com/EverestAPI/Everest/blob/dev/Celeste.Mod.mm/Mod/Everest/Everest.cs#L260)). And the env vars seem to match the build variables mentioned in https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml as well as conform to the format already present in the file. But otherwise as already mentioned idk what I'm doing lol

SnipUndercover commented:
You should be able to download the artifact from the Azure Pipelines site and install that.

Click the checkmark next to the commit in your PR, go to details, click the "3 artifacts" text, and finally download the `main` artifact. [Here's the direct download link.](https://dev.azure.com/EverestAPI/e877e744-b3c0-4e59-94d2-6042b8b565d2/_apis/build/builds/4859/artifacts?artifactName=main&api-version=7.1&%24format=zip)

You can install it from Olympus; scroll to the bottom of the Everest version list and pick the "Select from .zip" option, giving it the main.zip from AZP.

SnipUndercover reviewed (commented):
Looks good, but I'd like to test it myself first.

maddie480 commented:
> VersionCelesteString: 1.4.0.0-fna [Everest: 5559-azure-18f75-merge]

hm yes :stuck_out_tongue:

technically the pipeline runs on a "virtual merge" branch called `refs/pull/PULL_REQUEST_NUMBER/merge` and the docs explicitly say it takes the last path segment in the ref, so here we go

SnipUndercover commented on .azure-pipelines/prebuild.ps1:
```diff
@@ -1,7 +1,14 @@
 $EverestPath = [io.path]::combine($env:BUILD_SOURCESDIRECTORY, 'Celeste.Mod.mm', 'Mod', 'Everest', 'Everest.cs')
-(Get-Content $EverestPath) -replace '(?<=public readonly static string VersionString = ")[^"]*', "1.$BuildNumber.0-azure-$(($env:BUILD_SOURCEVERSION).Substring(0, 5))" | Set-Content $EverestPath
+# Get the branch name from Azure Pipelines
+$BranchName = $env:BUILD_SOURCEBRANCHNAME
```
```suggestion
if ($env:SYSTEM_PULLREQUEST_PULLREQUESTID -ne $null) {
  # Build triggered by a PR
  $BranchName = "pr$env:SYSTEM_PULLREQUEST_PULLREQUESTID"
} else {
  # Build triggered by a regular push
  $BranchName = $env:BUILD_SOURCEBRANCHNAME
}
```

I think knowing what PR this build belongs to would be neat - the [docs](https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml) mention a `System.PullRequest.PullRequestId` variable which is set to the PR number if the pipeline was triggered by a PR.

brtsai replied:
I had to read [this article](https://medium.com/@lizjohnson_83228/the-difference-between-push-and-pull-request-triggers-in-github-d7c0f2bc19db) to understand what you and Maddie were on about. So basically there are two general cases when the build pipeline builds. One when a PR is created to test the build for that PR and another when the changes are merged into the main repo. In this case the virtual merge branch is the branch used by the pipeline to build the changes for this PR and the branchname would be "refs/pull/918/merge".

This change would essentially make the version string in this case end in pr918 i.e.: "1.4.0.0-fna + 1.5509.0-azure-ed1e7-pr918".

Seems good and makes sense to me.

maddie480 replied:
Yeah, I learned about this about a week ago as well 😅 if you look at the latest pipeline run, you can see it's based on a commit that "does not exist": https://github.com/EverestAPI/Everest/commit/d5cb3b055de826656fc1cb365a8b60285fc3457e
And in our case, the takeaway is "the branch is going to be called `merge` on pull requests" 🤔

brtsai replied:
Ah. I can see that in the [build logs](https://dev.azure.com/EverestAPI/Everest/_build/results?buildId=4865&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=63e8042a-f367-5aec-692c-fcc654be55e4&l=511) for the azure pipeline.

Also I'm not sure why, but I can't seem to figure out how to download the published build artifacts to see what version string pops up for that olympus. Following the [azure devops docs](https://learn.microsoft.com/en-us/azure/devops/pipelines/artifacts/pipeline-artifacts?view=azure-devops&tabs=yaml#view-published-artifacts) I can view the published artifact list for the pipeline but I can't click through or expand any drop downs to download them.
<img width="272" alt="Screenshot 2025-06-11 at 6 08 28 AM" src="https://github.com/user-attachments/assets/03713397-74ea-4786-a75d-3c002fe250ed" />

SnipUndercover replied:
There's three dots at the very right side of the row.
Not sure why that's the design choice but it is what it is.

brtsai replied:
Ah I see. And they don't reveal until you hover the row.

SnipUndercover commented on .azure-pipelines/prebuild.ps1:
```diff
@@ -1,7 +1,14 @@
+# Get the branch name from Azure Pipelines
+$BranchName = $env:BUILD_SOURCEBRANCHNAME
+# Compose the version string with branch info
+$SourceVersionShort = $(($env:BUILD_SOURCEVERSION).Substring(0, 5))
```
```suggestion
$SourceVersionShort = $env:BUILD_SOURCEVERSION.Substring(0, 5)
```
The sub-expression syntax is no longer necessary.

brtsai replied:
thanks for the attention to detail!

SnipUndercover reviewed (changes requested):
Looked at this again and have some suggestions. If these could be implemented and turn out to work, I'll approve.

SnipUndercover commented:
A `VersionBranch` field in `Everest` similar to the ones linked below would also be useful, as I wanted some system in place for Everest to be able to warn regular end users when a development version is installed. The xmldocs should also be updated to reflect the new build format.
*Also, now that I look at it, the comment for `VersionCommit` is incorrect... it should say "commit" instead of "tag".*

https://github.com/EverestAPI/Everest/blob/cfa08a59bba69f7b0619f9f665fab562f2a59ea8/Celeste.Mod.mm/Mod/Everest/Everest.cs#L39-L58

maddie480 commented:
It doesn't quite look right... `5565-azure-d5cb3-pr2580697414`

maddie480 commented on .azure-pipelines/prebuild.ps1:
```diff
@@ -1,7 +1,20 @@
+# Get the branch name from Azure Pipelines
+if ($env:SYSTEM_PULLREQUEST_PULLREQUESTID -ne $null) {
+  # Build triggered by a PR
+  $BranchName = "pr$env:SYSTEM_PULLREQUEST_PULLREQUESTID"
```
```suggestion
  $BranchName = "pr$env:SYSTEM_PULLREQUEST_PULLREQUESTNUMBER"
```
That might be the one we want

brtsai replied:
Makes sense to me

maddie480 commented:
Looks better: `5566-azure-db5ee-pr918`

brtsai committed:
Include Branch [dev|stable|beta] or pr# at end of version string
Fixes #917

I.E.: "1.4.0.0-fna + 1.5509.0-azure-ed1e7" (old)
 => "1.4.0.0-fna + 1.5509.0-azure-ed1e7-dev" (new)
 => "1.4.0.0-fna + 1.5509.0-azure-ed1e7-pr918" (new)

SnipUndercover commented:
> A `VersionBranch` field in `Everest` similar to the ones linked below would also be useful, as I wanted some system in place for Everest to be able to warn regular end users when a development version is installed. The xmldocs should also be updated to reflect the new build format.
> 
> *Also, now that I look at it, the comment for `VersionCommit` is incorrect... it should say "commit" instead of "tag".*
> 
> 
> 
> https://github.com/EverestAPI/Everest/blob/cfa08a59bba69f7b0619f9f665fab562f2a59ea8/Celeste.Mod.mm/Mod/Everest/Everest.cs#L39-L58

Is this planned to be implemented in this PR or a future one?

brtsai commented:
> > A `VersionBranch` field in `Everest` similar to the ones linked below would also be useful, as I wanted some system in place for Everest to be able to warn regular end users when a development version is installed. The xmldocs should also be updated to reflect the new build format.
> > _Also, now that I look at it, the comment for `VersionCommit` is incorrect... it should say "commit" instead of "tag"._
> > https://github.com/EverestAPI/Everest/blob/cfa08a59bba69f7b0619f9f665fab562f2a59ea8/Celeste.Mod.mm/Mod/Everest/Everest.cs#L39-L58
> 
> Is this planned to be implemented in this PR or a future one?

Future one.

I referenced it in https://github.com/EverestAPI/Everest/issues/920 so it's documented. Because it's .cs application code and not pipeline .ps1 script I figured it was a different domain and the code change should be in a separate commit.

brtsai commented:
Is there a specific way to resolve requested changes to trigger the bot removing the "changes requested" label or does the bot just check periodically?
<img width="337" alt="image" src="https://github.com/user-attachments/assets/3949b747-2552-4019-a596-a2fa312e5ab9" />

Or did I just miss some requested changes?

maddie480 commented:
... it checks periodically. 😛 

About once an hour more specifically 😅
